maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelProcessor.java

```diff
@@ -81,7 +92,15 @@ public File locatePom(File projectDirectory) {
 
     public Path locatePom(Path projectDirectory) {
         // Note that the ModelProcessor#locatePom never returns null
-        Path pom = modelLocator.locatePom(projectDirectory.toFile()).toPath();
+        // while the ModelParser#locatePom needs to return an existing path!
+        Path pom = modelParsers.stream()
+                .map(m -> m.locate(projectDirectory)
+                        .map(org.apache.maven.api.services.Source::getPath)
+                        .orElse(null))
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElseGet(
+                        () -> modelLocator.locatePom(projectDirectory.toFile()).toPath());
         if (!pom.equals(projectDirectory) && !pom.getParent().equals(projectDirectory)) {
             throw new IllegalArgumentException("The POM found does not belong to the given directory: " + pom);
```

Comment:
I'm not sure this is the right exception here. This isn't necessarily a problem with he argument itself. It's more akin to a checked IOException, again arising from conditions external to the program itself.

Reply:
I strongly disagree.  This is a health check on the response. Both `ModelLocator#locatePom` and `ModelParser#locate` are supposed to return a file in the provided directory or the directory itself.  This is quite explicit in their respective javadoc.  So it's a programming error and can not be recovered.